deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/AsyncDataSetIterator.java

```diff
@@ -113,291 +93,282 @@ public AsyncDataSetIterator(DataSetIterator iterator, int queueSize, BlockingQue
-            if (!runnable.killRunnable && runnable.exception != null) {
-                throw runnable.exception; //Something went wrong
-            }
-            //Runnable has exited, presumably because it has fetched all elements
-            return runnable.hasLatch();
-        }
-    }
+        try {
+            if (nextElement != null && nextElement != terminator) {
+                return true;
+            } else if(nextElement == terminator)
+                return false;
 
-    @Override
-    public synchronized DataSet next() {
-        if (!hasNext()) {
-            throw new NoSuchElementException();
-        }
-        //If base iterator threw an unchecked exception: rethrow it now
-        if (runnable.exception != null) {
-            throw runnable.exception;
-        }
 
-        if (!blockingQueue.isEmpty()) {
-            runnable.feeder.decrementAndGet();
-            try {
-                return blockingQueue.poll(2, TimeUnit.SECONDS); //non-blocking, but returns null if empty
-            } catch (InterruptedException e) {
-                //
-            }
-        }
+            nextElement = buffer.take();
 
-        //Blocking queue is empty, but more to come
-        //Possible reasons:
-        // (a) runnable died (already handled - runnable.exception != null)
-        // (b) baseIterator.next() hasn't returned yet -> wait for it
-        try {
-            //Normally: just do blockingQueue.take(), but can't do that here
-            //Reason: what if baseIterator.next() throws an exception after
-            // blockingQueue.take() is called? In this case, next() will never return
-            while (runnable.exception == null) {
-                DataSet ds = blockingQueue.poll(2, TimeUnit.SECONDS);
-                if (ds != null) {
-                    runnable.feeder.decrementAndGet();
-                    return ds;
-                }
-                if (runnable.killRunnable) {
-                    //should never happen
-                    throw new ConcurrentModificationException("Reset while next() is waiting for element?");
-                }
-                if (!runnable.isAlive.get() && blockingQueue.isEmpty()) {
-                    if (runnable.exception != null)
-                        throw new RuntimeException("Exception thrown in base iterator", runnable.exception);
-                    throw new IllegalStateException(
-                            "Unexpected state occurred for AsyncDataSetIterator: runnable died or no data available");
-                }
-            }
-            //exception thrown while getting data from base iterator
-            throw runnable.exception;
-        } catch (InterruptedException e) {
-            throw new RuntimeException(e); //Shouldn't happen under normal circumstances
+            if (nextElement == terminator)
+                return false;
+
+            return true;
+        } catch (Exception e) {
+            log.error("Premature end of loop!");
```

Comment:
There shouldn't be any besides of InterruptedException, really. It's only buffer.take() happens here.
And real exceptions are propagated from prefetch thread.
However, let's just add it.